libbeat/common/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,48 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package diagnostics
+
+// DiagnosticRunner is a sub-type of the runner interface found as cfgfile.Runner
+// Runners can implement this if they want their client filesets/metricsets to expose the underlying DiagnosticSet interface
+type DiagnosticRunner interface {
+	ModuleDiagnostics() []DiagnosticSetup
+}
+
+// DiagnosticSet is an interface that a metricset/fileset can implement to provide additional Diagnostic data.
+// A metricset can provide any number of diagnostic responses when requested.
+type DiagnosticSet interface {
+	// DiagnosticSetup returns metadata and a callback halder.
+	// note that this can be called any time after a metricset has started, so implementors should not assume
+	// the state of a metricset/fileset when this method is called.
+	DiagnosticSetup() []DiagnosticSetup
+}
+
+// DiagnosticSetup contains the data needed to register a callback.
+type DiagnosticSetup struct {
+	// The name of this diagnostics data result.
+	Name string
```

Comment:
Is `Name` used for anything? It seems like we might just need `Filename`

Reply:
So, it's used as a field in the client `RegisterDiagnosticHook()`, and I wanted this type to be a somewhat un-opinionated wrapper for that.